lighthouse-core/test/audits/users-rel-preconnect-test.js

```diff
@@ -0,0 +1,206 @@
+      mainResourceRecord,
+      {
+        url: 'https://cdn.example.com/request',
+        initiatorRequest: () => null,
+        _startTime: 16,
+      },
+    ];
+    const artifacts = {
+      devtoolsLogs: {[UsesRelPreconnect.DEFAULT_PASS]: []},
+      requestNetworkRecords: () => Promise.resolve(networkRecords),
+      requestMainResource: () => Promise.resolve(mainResource),
+    };
+
+    const {score, rawValue, details} = await UsesRelPreconnect.audit(artifacts);
+    assert.equal(score, 1);
+    assert.equal(rawValue, 0);
+    assert.equal(details.items.length, 0);
+  });
+
+  it(`should only list an origin once`, async () => {
+    const networkRecords = [
+      mainResourceRecord,
+      {
+        url: 'https://cdn.example.com/first',
+        initiatorRequest: () => null,
+        _startTime: 2,
+        timing: {
+          dnsStart: 100,
+          connectStart: 150,
+          connectEnd: 300,
+        },
+      },
+      {
+        url: 'https://cdn.example.com/second',
+        initiatorRequest: () => null,
+        _startTime: 3,
+        timing: {
+          dnsStart: 300,
+          connectStart: 350,
+          connectEnd: 400,
+        },
+      },
+    ];
+    const artifacts = {
+      devtoolsLogs: {[UsesRelPreconnect.DEFAULT_PASS]: []},
+      requestNetworkRecords: () => Promise.resolve(networkRecords),
+      requestMainResource: () => Promise.resolve(mainResource),
+    };
+
+    const {rawValue, extendedInfo} = await UsesRelPreconnect.audit(artifacts);
+    assert.equal(rawValue, 200);
+    assert.equal(extendedInfo.value.length, 1);
+    assert.deepStrictEqual(extendedInfo.value, [
+      {url: 'https://cdn.example.com', wastedMs: Util.formatMilliseconds(200)},
+    ]);
+  });
+
+  it(`should give a list of preconnected origins`, async () => {
+    const networkRecords = [
+      mainResourceRecord,
+      {
+        url: 'https://cdn.example.com/first',
+        initiatorRequest: () => null,
+        _startTime: 2,
+        timing: {
+          dnsStart: 100,
+          connectStart: 250,
+          connectEnd: 300,
+        },
+      },
+      {
+        url: 'https://othercdn.example.com/second',
+        initiatorRequest: () => null,
+        _startTime: 1.2,
+        timing: {
+          dnsStart: 100,
+          connectStart: 200,
+          connectEnd: 600,
+        },
+      },
+    ];
+    const artifacts = {
+      devtoolsLogs: {[UsesRelPreconnect.DEFAULT_PASS]: []},
+      requestNetworkRecords: () => Promise.resolve(networkRecords),
+      requestMainResource: () => Promise.resolve(mainResource),
+    };
+
+    const {rawValue, extendedInfo} = await UsesRelPreconnect.audit(artifacts);
+    assert.equal(rawValue, 400);
+    assert.equal(extendedInfo.value.length, 2);
+    assert.deepStrictEqual(extendedInfo.value, [
+      {url: 'https://cdn.example.com', wastedMs: Util.formatMilliseconds(200)},
+      {url: 'https://othercdn.example.com', wastedMs: Util.formatMilliseconds(400)},
```

Comment:
would you mind explaining this result? seems like it should be min of `500 (600 - 100)` and `300 (1000 * (1.2 - 1) + 100)`

Reply:
when looking in code I use connectStart where I should have used dnsStart 🤦‍♂️ 
```
const connectionTime = 600 - 100;
const timeBetweenMainResourceAndConnectStart =
  1.2 * 1000 -
   1 * 1000 +
  200;
// calculate delta between connectionTime and timeToConnectionStart from main resource
const wastedMs = Math.min(500, 400);
```

Reply:
ah yeah I missed that too :) seems like `timeBetweenMainResourceAndDnsStart` should fix it 👍 (or fallback to connect start I guess if dns start is invalid, is that a real case?)